lib/reporting/reports/sales_tax/sales_tax_totals_by_producer.rb

```diff
@@ -0,0 +1,169 @@
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/ClassLength
+module Reporting
+  module Reports
+    module SalesTax
+      class SalesTaxTotalsByProducer < Base
+        def search
+          report_line_items.orders
+        end
+
+        def order_permissions
+          return @order_permissions unless @order_permissions.nil?
+
+          @order_permissions = ::Permissions::Order.new(user, ransack_params)
+        end
+
+        def report_line_items
+          @report_line_items ||= Reporting::LineItems.new(order_permissions, params)
+        end
+
+        def query_result
+          # The objective is to group the line items by
+          # [tax_rate, supplier_id, distributor_id and order_cycle_id]
+          # As every line item can match more than one tax_rate
+          # We'll do it in two steps
+          # 1. we group by [[tax_rate], supplier_id, distributor_id , order_cycle_id]
+          # 2. we split each key that has more than tax_rate
+          report_line_items.list.group_by do |line_item|
+            [
+              line_item.tax_rates.map(&:id),
+              line_item.supplier_id,
+              line_item.order.distributor_id,
+              line_item.order.order_cycle_id
+            ]
+          end.flat_map do |k, v|
+            k.first.map do |tax_rate_id|
+              {
+                [tax_rate_id] + k[1..] => v
+              }
+            end
+          end
+        end
+
+        def columns
+          {
+            distributor: :distributor,
+            distributor_tax_status: :distributor_tax_status,
+            producer: :producer,
+            producer_tax_status: :producer_tax_status,
+            order_cycle: :order_cycle,
+            tax_category: :tax_category,
+            tax_rate_name: :tax_rate_name,
+            tax_rate: :tax_rate_amount,
+            total_excl_tax: :total_excl_tax,
+            tax: :tax,
+            total_incl_tax: :total_incl_tax
+          }
+        end
+
+        def rules
+          [
+            {
+              group_by: :distributor,
```

Comment:
Do we need the report with multiple distributors? We could enforce a selection in the UI and then have only one distributor per report. This limitation could improve performance.

Reply:
> Do we need the report with multiple distributors?

I think @lin-d-hop  can answer this question : )

Reply:
@mkllnk This would introduce inconsistency for this report. Is performance already proving a problem or are you speculating?